tools/clang/test/CodeGenSPIRV/rich.debug.member.function.param.hlsl

```diff
@@ -0,0 +1,42 @@
+// Run: %dxc -T ps_6_0 -E main -fspv-debug=rich
+
+struct foo {
+  int a;
+
+  void func0(float arg) {
+    b.x = arg;
+  }
+
+  float4 b;
+
+  int func1(int arg0, float arg1, bool arg2) {
+    a = arg0;
+    b.y = arg1;
+    if (arg2) return arg0;
+    return b.z;
+  }
+
+  bool c;
+};
+
+// CHECK: [[func0:%\d+]] = OpString "foo.func0"
+// CHECK: [[arg:%\d+]] = OpString "arg"
+// CHECK: [[func1:%\d+]] = OpString "foo.func1"
+// CHECK: [[arg0:%\d+]] = OpString "arg0"
+// CHECK: [[arg1:%\d+]] = OpString "arg1"
+// CHECK: [[arg2:%\d+]] = OpString "arg2"
+
+// CHECK: [[f0]] = OpExtInst %void [[set]] DebugFunction [[func0]] {{%\d+}} {{%\d+}} 6 3 [[parent]] {{%\d+}} FlagIsProtected|FlagIsPrivate 6 %foo_func0
+// CHECK: {{%\d+}} = OpExtInst %void [[set]] DebugLocalVariable [[arg]] {{%\d+}} {{%\d+}} 6 20 [[f0]] FlagIsLocal 0
+// CHECK: [[f1]] = OpExtInst %void [[set]] DebugFunction [[func1]] {{%\d+}} {{%\d+}} 12 3 [[parent]] {{%\d+}} FlagIsProtected|FlagIsPrivate 12 %foo_func1
+// CHECK: {{%\d+}} = OpExtInst %void [[set]] DebugLocalVariable [[arg0]] {{%\d+}} {{%\d+}} 12 17 [[f1]] FlagIsLocal 0
+// CHECK: {{%\d+}} = OpExtInst %void [[set]] DebugLocalVariable [[arg1]] {{%\d+}} {{%\d+}} 12 29 [[f1]] FlagIsLocal 1
+// CHECK: {{%\d+}} = OpExtInst %void [[set]] DebugLocalVariable [[arg2]] {{%\d+}} {{%\d+}} 12 40 [[f1]] FlagIsLocal 2
```

Comment:
I mentioned this in another comment about parameter indexes being off by 1:

The above `DebugFunction` instruction is associated with `%foo_func1`.

If you look at the SPIR-V, you'll see the signature fore `%foo_func1` is:
```spir-v
  %foo_func1 = OpFunction %int None %51
%param_this_0 = OpFunctionParameter %_ptr_Function_foo
       %arg0 = OpFunctionParameter %_ptr_Function_int
       %arg1 = OpFunctionParameter %_ptr_Function_float
       %arg2 = OpFunctionParameter %_ptr_Function_bool
```

Meaning, the first parameter is the `this` object (of type `foo`). If we want `DebugLocalVariable` to match the signature of `%foo_func1`, then the parameter indexes should be shifted by one:

```spir-v
// CHECK: {{%\d+}} = OpExtInst %void [[set]] DebugLocalVariable [[arg0]] {{%\d+}} {{%\d+}} 12 17 [[f1]] FlagIsLocal 1
// CHECK: {{%\d+}} = OpExtInst %void [[set]] DebugLocalVariable [[arg1]] {{%\d+}} {{%\d+}} 12 29 [[f1]] FlagIsLocal 2
// CHECK: {{%\d+}} = OpExtInst %void [[set]] DebugLocalVariable [[arg2]] {{%\d+}} {{%\d+}} 12 40 [[f1]] FlagIsLocal 3
```

I don't know how the debuggers handle class/struct member functions. If they expect to match with the HLSL signature, then your existing code is correct. If they expect to match the SPIR-V signature, then you need to shift the parameter indexes in `DebugLocalVariable` by 1. What do you think?

Reply:
I am suddenly worried about this as well. I just opened a bug for this issue. Let's discuss it there.

Reply:
also seems like the flag needs to be updated.